sci/ramfuzz.md

```diff
@@ -0,0 +1,102 @@
+
+- All values are converted to a single type, so they can be packed into a NumPy array.  The type is `np.float64`, to ensure the maximum possible range for receiving logged values.  Location indices are of type `np.uint64` for the same reasons.
+
+Because the ranges of generated values can be extremely wide, the original model's training was not effective: its weights would reach extreme values or NaNs, resulting in low accuracy.  We introduced a batch-normalization layer to reduce the values before any further processing.  We also had to significantly lower the batch size to avoid NaNs in batch-normalizer's mean calculation.  The resulting model is also [available on GitHub](https://github.com/dekimir/RamFuzz/blob/master/ai/sample-model1.py).
+
+After these preparations, we were able to train the model to predict whether a test was successful or not from the contents of its log.  The training achieved over 99% prediction accuracy within five epochs.  These results were consistent across multiple runs on both Linux and Mac OS.  In one Linux run, the training actually achieved 100% accuracy.
+
+## Conclusions
+
+The RamFuzz code generator makes it easy to create randomized unit tests in C++, and its AI tools make it easy to generate unlimited quantities of training data and feed it to deep-learning models.  RamFuzz will therefore enable more effective research in several important directions.
+
+The surprising accuracy of deep learning at predicting test failures suggests that AI can play a major role in automated test generation.  Consider that the trained deep-learning model in fact represents AI's knowledge of the tested software's specification.  In time, we may be able to translate such specifications into human language, check them for code quality, or even compare specifications of different programs (or different versions of the same program).  More ambitiously, we may be able to use the trained model's knowledge to constrain the runtime generation of random values in a way that yields only valid tests.  This would give us the first autonomous system for testing existing C++ software.
+
+One notable shortcoming of RamFuzz currently is that it falls short of generating entire unit tests.  It generates test inputs but not test assertions.  In fact, this is common with many other tools for automatic test generation, and it stems from the simple fact that the process is missing another input: the source code's specification.  Without a specification, it is simply impossible to devise tests that verify whether the code conforms to it.  Nevertheless, the experience of prior tools for autonomous test generation is that even tests with trivial or missing assertions can be useful in finding bugs [2, 3].
```

Comment:
I find the assertion here that randomly generated tests can be helpful in finding real world bugs tantalizing, but without having the time at the moment to go look at those references, I'm not immediately clear on how that would work. It might be helpful to recount an example or two here of how that could play out, to provide a little better substantiation for the value of the overall approach.

Reply:
Thanks; I'll add something like "In one case study, Randoop found 30 new bugs in Microsoft production code that had previously undergone 200 man years of testing; sample bugs ranged from memory corruption to missing messages in resource files."

https://randoop.github.io/randoop/files/randoop_case_study_2008.pdf

Reply:
Done.